// seamless-scroll-polyfill/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.seamless = exports.windowScrollToPolyfill = exports.windowScrollTo = exports.windowScrollByPolyfill = exports.windowScrollBy = exports.windowScrollPolyfill = exports.windowScroll = exports.elementScrollToPolyfill = exports.elementScrollTo = exports.elementScrollIntoViewPolyfill = exports.elementScrollIntoView = exports.elementScrollByPolyfill = exports.elementScrollBy = exports.elementScrollPolyfill = exports.elementScroll = exports.polyfill = void 0;
const common_js_1 = require("./common.js");
const Element_scroll_js_1 = require("./Element.scroll.js");
const Element_scrollBy_js_1 = require("./Element.scrollBy.js");
const Element_scrollIntoView_js_1 = require("./Element.scrollIntoView.js");
const Element_scrollTo_js_1 = require("./Element.scrollTo.js");
const Window_scroll_js_1 = require("./Window.scroll.js");
const Window_scrollBy_js_1 = require("./Window.scrollBy.js");
const Window_scrollTo_js_1 = require("./Window.scrollTo.js");
exports.polyfill = (options) => {
    if (common_js_1.isScrollBehaviorSupported()) {
        return;
    }
    Window_scroll_js_1.windowScrollPolyfill(options);
    Window_scrollTo_js_1.windowScrollToPolyfill(options);
    Window_scrollBy_js_1.windowScrollByPolyfill(options);
    Element_scroll_js_1.elementScrollPolyfill(options);
    Element_scrollTo_js_1.elementScrollToPolyfill(options);
    Element_scrollBy_js_1.elementScrollByPolyfill(options);
    Element_scrollIntoView_js_1.elementScrollIntoViewPolyfill(options);
};
exports.seamless = exports.polyfill;
var Element_scroll_js_2 = require("./Element.scroll.js");
Object.defineProperty(exports, "elementScroll", { enumerable: true, get: function () { return Element_scroll_js_2.elementScroll; } });
Object.defineProperty(exports, "elementScrollPolyfill", { enumerable: true, get: function () { return Element_scroll_js_2.elementScrollPolyfill; } });
var Element_scrollBy_js_2 = require("./Element.scrollBy.js");
Object.defineProperty(exports, "elementScrollBy", { enumerable: true, get: function () { return Element_scrollBy_js_2.elementScrollBy; } });
Object.defineProperty(exports, "elementScrollByPolyfill", { enumerable: true, get: function () { return Element_scrollBy_js_2.elementScrollByPolyfill; } });
var Element_scrollIntoView_js_2 = require("./Element.scrollIntoView.js");
Object.defineProperty(exports, "elementScrollIntoView", { enumerable: true, get: function () { return Element_scrollIntoView_js_2.elementScrollIntoView; } });
Object.defineProperty(exports, "elementScrollIntoViewPolyfill", { enumerable: true, get: function () { return Element_scrollIntoView_js_2.elementScrollIntoViewPolyfill; } });
var Element_scrollTo_js_2 = require("./Element.scrollTo.js");
Object.defineProperty(exports, "elementScrollTo", { enumerable: true, get: function () { return Element_scrollTo_js_2.elementScrollTo; } });
Object.defineProperty(exports, "elementScrollToPolyfill", { enumerable: true, get: function () { return Element_scrollTo_js_2.elementScrollToPolyfill; } });
var Window_scroll_js_2 = require("./Window.scroll.js");
Object.defineProperty(exports, "windowScroll", { enumerable: true, get: function () { return Window_scroll_js_2.windowScroll; } });
Object.defineProperty(exports, "windowScrollPolyfill", { enumerable: true, get: function () { return Window_scroll_js_2.windowScrollPolyfill; } });
var Window_scrollBy_js_2 = require("./Window.scrollBy.js");
Object.defineProperty(exports, "windowScrollBy", { enumerable: true, get: function () { return Window_scrollBy_js_2.windowScrollBy; } });
Object.defineProperty(exports, "windowScrollByPolyfill", { enumerable: true, get: function () { return Window_scrollBy_js_2.windowScrollByPolyfill; } });
var Window_scrollTo_js_2 = require("./Window.scrollTo.js");
Object.defineProperty(exports, "windowScrollTo", { enumerable: true, get: function () { return Window_scrollTo_js_2.windowScrollTo; } });
Object.defineProperty(exports, "windowScrollToPolyfill", { enumerable: true, get: function () { return Window_scrollTo_js_2.windowScrollToPolyfill; } });
